### PR TITLE
Isolate determining the encoding type into its own function.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libsolidity/ast/Types.cpp [hunk 1]
```diff
@@ -390,6 +390,27 @@ MemberList const& Type::members(ContractDefinition const* _currentScope) const
...
 }
 
+TypePointer Type::fullEncodingType(bool _inLibraryCall, bool _encoderV2, bool _packed) const
+{
+	TypePointer encodingType = mobileType();
+	if (encodingType)
+		encodingType = encodingType->interfaceType(_inLibraryCall);
+	if (encodingType)
+		encodingType = encodingType->encodingType();
+	if (auto structType = dynamic_cast<StructType const*>(encodingType.get()))
```
Comment (axic) [inline libsolidity/ast/Types.cpp:400]: ~~If above `encodingType` is null how will `.get()` work?~~ Ah it's a shared pointer.

**Comment by axic** [comment]
Rebased this to see if it breaks any tests added since.

**Review by axic — APPROVED** [review]

**Comment by axic** [comment]
@ekpyron do you also want to have a look?

File: libsolidity/codegen/CompilerUtils.cpp [hunk 2]
```diff
@@ -333,26 +333,19 @@ void CompilerUtils::encodeToMemory(
...
 	for (TypePointer& t: targetTypes)
 	{
-		solUnimplementedAssert(
-			t->mobileType() &&
-			t->mobileType()->interfaceType(_encodeAsLibraryTypes) &&
-			t->mobileType()->interfaceType(_encodeAsLibraryTypes)->encodingType(),
-			"Encoding type \"" + t->toString() + "\" not yet implemented."
-		);
-		t = t->mobileType()->interfaceType(_encodeAsLibraryTypes)->encodingType();
+		TypePointer tEncoding = t->fullEncodingType(_encodeAsLibraryTypes, encoderV2, !_padToWordBoundaries);
```
Comment (ekpyron) [inline libsolidity/codegen/CompilerUtils.cpp:341]: What purpose does ``tEncoding`` serve? Why not directly move into ``t`` and assert ``t`` instead below?

**Review by ekpyron — COMMENTED** [review]
Apart from the minor comment this looks fine.

File: libsolidity/codegen/CompilerUtils.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/codegen/CompilerUtils.cpp:341]: If we overwrite `t`, we cannot use it in the assertion message...

Comment (ekpyron) [reply libsolidity/codegen/CompilerUtils.cpp:341]: Ah, I hadn't seen that it's used there.

**Review by ekpyron — APPROVED** [review]

**Pull request merged**